### PR TITLE
Packaging for PR #2719 to promote JupyterHub programming with Notebooks

### PR DESCRIPTION
**Pull request by holta**
This PR was tested on Ubuntu Server 21.04 Beta: (VM)

1) Packages @georgejhunt's PR #2719 to be promoted more widely as part of BIG-sized IIAB installs.
2) Installs JupyterHub in roles/6-generic-apps/tasks/main.yml (right after Gitea, which likewise aims to serve high school programming and similar maker communities).
3) Fixes a regression in `roles/network/templates/gateway/iiab-gen-iptables` where `$transmission_peer_port` was accidentally being lost: https://github.com/iiab/iiab/pull/2719/files#diff-ff08907ce43e8e1e073a32b255c8aa8ba883ed1e12e425314e435d097e86c52cR163
4) Cleans up documentation in many places, alongside http://FAQ.IIAB.IO -> "9 What services (IIAB apps) are suggested during installation?"

_Educators and Implementers should continue to enhance_ https://github.com/iiab/iiab/blob/master/roles/jupyterhub/README.md _towards seeding a supportive & growing community here!_

### CODE REVIEW

**Comment by holta** [comment]
Reinstalled & re-tested on Ubuntu Server 21.04 VM for further validation:

- What user-facing testing is needed at http://10.8.0.62/jupyterhub, if any?
- What's the recommended way for educators to quickly set up an account and explore this?

File: roles/jupyterhub/tasks/install.yml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+- name: Use npm to install configurable-http-proxy
+  npm:
+    name: configurable-http-proxy
+    global: yes
+    state: latest
+
+- name: 'Use pip to install into a virtual environment: {{ jupyterhub_venv }}'
+  pip:
+    name:
+      - pip
```
Comment (holta) [inline roles/jupyterhub/tasks/install.yml:34]: Interesting that pip is installing pip (as circular as that appears!)

@georgejhunt do you happen to know how JupyterHub uses this?

Total Aside: pip3 is often installed by apt package `python3-pip` in 2021.

**Comment by holta** [comment]
FYI I've added this description to https://github.com/iiab/iiab/wiki/IIAB-7.2-Release-Notes#whats-new :

> JupyterHub Notebooks [https://en.wikipedia.org/wiki/Project_Jupyter] (README [https://github.com/iiab/iiab/blob/master/roles/jupyterhub/README.md]) give schools a new way to integrate programming projects, allowing students to integrate science experiment results (e.g. dynamic interactive graphing) with their program output, in a "Notebook" document or blog.

**Comment by holta** [comment]
FYI this was also successfully installed on 32-bit RaspiOS Lite on RPi 4, as part of a BIG-sized IIAB install here:

http://10.8.0.38/jupyterhub

**Caution: this particular test machine is intentionally underpowered, with only 1GB RAM, for purposes of stress-testing!**

**Comment by holta** [comment]
@georgejhunt can you explain the sentence below from https://github.com/holta/iiab/blob/jupyter/roles/jupyterhub/README.md ?

> This IIAB package permits individual users to start using their own notebook on the server (http://box.lan/jupyterhub) without needing an individual server account.

Specifically, how does an educator proceed to make this happen &mdash; with or without creating account(s) ?

Somehow we need to make it Dead Easy for educators & implementers to explore this quickly / efficiently / without getting lost, Thanks!  (Currently it's not at all obvious what they would + should do when facing things like http://10.8.0.38/jupyterhub and http://10.8.0.62/jupyterhub &mdash; _what tips get them moving ?_ )

File: roles/jupyterhub/tasks/install.yml (hunk 1 above)
Comment (jvonau) [reply roles/jupyterhub/tasks/install.yml:34]: That is a pretty snide comment, guess you just misunderstood that pip is installing needed dependencies into a blank isolated virtual environment where there is nothing, apart from where python3-pip lives. Can't recall off the top of my head if pip is auto-installed when the virtual environment is created or not but PyPI should install all the declared dependencies   but just being explicit is a0k in my books also. The other way would be to reference the project's requirements.txt file if one exists with a github source(calibre-web).

**Comment by jvonau** [comment]
With http://host/jupyterhub working there really is no need to open the port in the firewall, and there was a dup'd stanza in iiab-gen-iptables.but that was fixed

**Comment by georgejhunt** [comment]
When I tried to generate a simple first "hello world" demo, the javascript
client fails to connect with the single-user python kernel. That's a bug,
and it may take a while for me to fix. But it's now my top priority

**Comment by holta** [comment]
> When I tried to generate a simple first "hello world" demo, the javascript client fails to connect with the single-user python kernel. That's a bug, and it may take a while for me to fix. But it's now my top priority

Thanks!  I'm merging this into master as we're now very close.  As this converges, let's:

1) Nail the above bug that @georgejhunt is working on.
2) Clarify & resolve "no need to open the port in the firewall" that @jvonau mentioned.
3) Test with a fresh install (final retesting) on both Ubuntu 21.04 and RaspiOS.
4) Please throw in a couple more on-ramping / acnt-creation tips (or link to any Jupyter Community "cheat sheets" ?) within https://github.com/iiab/iiab/blob/master/roles/jupyterhub/README.md so eager educators can **hit the ground running!**

**Comment by holta** [comment]
FYI this PR is further revised by:

- PR #2736 "Jhubfix"
- PR #2737 "Lint jupyterhub-nginx.conf & nginx.conf"

**Pull request merged**